[Merged by Bors] - Perform invlinking in assume rather than implicitly in getindex

torfjelde opened the pull request:
Currently, in `assume`, etc., `invlink` is called implicitly in `getindex` using the distribution extracted from `vi`.

This has a couple of drawbacks:
1. We can only use the distribution for a particular `vn` stored in `vi` obtained during the initial run. This means that we can't even run models where the distributions has dynamic domains, i.e. the domain of a particular random variable is dependent on the realizations of other random variables.
2. We have to store the distribution for each `vn` in `vi`. This was fine when we only had `VarInfo` because we also need it for other functionality, but this is not the case in `SimpleVarInfo` (nor will it be).

So. In this PR we introduce a `getindex_raw` which is `getindex` but without `invlink` if it's already linked, and uses this within `assume`, etc. where we now use the distributions that are passed to `assume` rather than those stored in `vi`.

E.g. the following now works:

``` julia
julia> @model demo() = x ~ InverseGamma(2, 3)
demo (generic function with 2 methods)

julia> vi = SimpleVarInfo((x = 10.0, ), true)
SimpleVarInfo((x = 10.0,), 0.0, true)

julia> _, vi = DynamicPPL.evaluate!!(model, vi, DefaultContext())
(22026.465794806718, SimpleVarInfo{NamedTuple{(:x,), Tuple{Float64}}, Float64}((x = 10.0,), -17.80291162245307, true))
```

devmotion commented:
Maybe `getindex` should be `getindex_raw` and indexing with linking should use a different notation?

torfjelde commented:
> Maybe `getindex` should be `getindex_raw` and indexing with linking should use a different notation?

Ideally this is indeed how we do it. The current impl is super-confusing, e.g. if you use a sampler in `getindex` it will not transform the variables, and `setindex` expects values to be in constrained space, irregardless of whether you're indexing using a `VarName` or `Sampler`.

_But_ I assume we need to do a bit of tracking down of all references to `vi[vn]` and ensure that we're not breaking anything in downstream packages, etc. Therefore my intention was to first just get a quick and dirty impl in that addresses the points mentioned in this PR, and then we make the indexing behavior consistent in a separate one.

torfjelde commented on src/simple_varinfo.jl:
```diff
@@ -129,11 +129,20 @@ ERROR: type NamedTuple has no field b
 struct SimpleVarInfo{NT,T} <: AbstractVarInfo
     values::NT
     logp::T
+    # TODO: Should we put this in the type instead?
```
Any thoughts on this?

torfjelde commented:
bors try

torfjelde commented:
bors try

torfjelde commented:
Tests are failing on Julia 1.3 because the fix from https://github.com/TuringLang/Turing.jl/pull/1758 hasn't made it through yet.

(It doesn't fail on Julia#latest because it's Julia 1.7 which makes it incompatible with latest release of Turing)

yebai commented:
@torfjelde it’s ok to make a new release from `Turing#master`.

torfjelde commented:
Hah, yeah I just did 10sec after your comment. Noticed you had reverted the changes from a couple of days ago; thanks!

torfjelde commented:
But this PR should be read for a round of reviews now then.

The main question is whether we want to just make `getindex_raw` be `getindex`, making it a breaking change, or if we want to keep `getindex_raw` for now, make a non-breaking change, and then we make a breaking change at a later stage. 

IMO we do the former, since it means that we can make use of `SimpleVarInfo` for models requiring linking _now_ + `getindex_raw` is only used internally.

torfjelde commented:
bors try

torfjelde commented:
bors try

devmotion commented on src/DynamicPPL.jl:
```diff
@@ -176,4 +176,8 @@ include("test_utils.jl")
 @deprecate acclogp!(vi, logp) acclogp!!(vi, logp)
 @deprecate resetlogp!(vi) resetlogp!!(vi)
 
+@deprecate settrans!(vi::AbstractVarInfo, trans::Bool, vn::VarName) settrans!!(
+    vi, trans, vn
+)
```
This exports `settrans!!`, is this intended?
```suggestion
@deprecate settrans!(vi::AbstractVarInfo, trans::Bool, vn::VarName) settrans!!(
    vi, trans, vn
) false
```

torfjelde replied:
Not intended, but good question: do we want to do this? If we export `SimpleVarInfo` it might make sense to also export `settrans!!` since it's a pretty central functionality of working with a varinfo instance.

yebai replied:
If it is not yet exported, I suggest we keep it as-is. We probably want to replace `settrans!!` with a more unified design with other `VarInfo` metadata such as `num_produce`, `is_deleted` etc.

devmotion commented on src/context_implementations.jl:
```diff
@@ -462,19 +474,24 @@ function get_and_set_val!(
-            settrans!(vi, false, vn)
+            if istrans(vi)
+                push!!(vi, vn, maybe_link(vi, vn, dist, r[:, i]), dist, spl)
+                # `push!!` sets the trans-flag to `false` by default.
+                setttrans!!(vi, true, vn)
```
This seems to be another confusing design choice but probably should not be addressed in this PR.

torfjelde replied:
See my comment above.

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -126,14 +126,35 @@ ERROR: type NamedTuple has no field b
-SimpleVarInfo{T}(θ) where {T<:Real} = SimpleVarInfo{typeof(θ),T}(θ, zero(T))
-SimpleVarInfo{T}(; kwargs...) where {T<:Real} = SimpleVarInfo{T}(NamedTuple(kwargs))
-SimpleVarInfo(; kwargs...) = SimpleVarInfo{Float64}(NamedTuple(kwargs))
+function Setfield.ConstructionBase.constructorof(
```
Maybe cleaner to depend on ConstructionBase directly? And is this the first time we use it?

torfjelde replied:
I'm fine with doing that, yeah.

Yes it's the first time using it explicitly like this (but it's fundamental to impl of Setfield.jl).

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -126,14 +126,35 @@ ERROR: type NamedTuple has no field b
+    end
+end
+
+function SimpleVarInfo(values, logp, istrans::Bool=false)
+    return SimpleVarInfo(values, logp, Val{istrans}())
+end
+function SimpleVarInfo(values, logp, ::Val{IsTrans}) where {IsTrans}
+    return SimpleVarInfo{typeof(values),typeof(logp),IsTrans}(values, logp)
+end
+
+function SimpleVarInfo{T}(θ) where {T<:Real}
+    return SimpleVarInfo{typeof(θ),T,false}(θ, zero(T))
+end
+function SimpleVarInfo{T}(; kwargs...) where {T<:Real}
+    return SimpleVarInfo{T}(NamedTuple(kwargs))
+end
+function SimpleVarInfo(; kwargs...)
+    return SimpleVarInfo{Float64}(NamedTuple(kwargs))
+end
```
Do we need all of them?

torfjelde replied:
I was also thinking all these constructors are maybe a bit too much :sweat_smile: 

_But_ I think they're all useful:
1. `SimpleVarInfo(θ)` or `SimpleVarInfo(; kwargs...)` is probably the most convenient (and will be most used) , and so I think these are worth having around.
2. `SimpleVarInfo(values, logp)` is also very convenient as it allows specific setting of the `logp` field.
3. But we also want ways to set `istrans`, hence we want the `SimpleVarInfo(values, logp, istrans)`.

_But_ now that I'm thinking about it, maybe we can  just drop the `{T}` constructors since we can equivalently just call `SimpleVarInfo(θ, zero(T))`? Only thing we lose is the ability to use the keyword-constructor while simultaneously specifying the type of `logp`, but maybe that's fine.

Whadda yah think? I'm fine with dropping the `{T}` constructors and maybe dropping the keyword-based one (though it looks real nice).

devmotion replied:
I think it would be nice to replace the `SimpleVarInfo(values, logp, ::Val)` constructor with `SimpleVarInfo{IsTrans}(values, logp)`, this seems cleaner than `Val`. I guess this means that probably we should remove the existing `SimpleVarInfo{T}` constructors since otherwise it becomes confusing.

So maybe:
- `SimpleVarInfo{IsTrans}(values, logp)`
- `SimpleVarInfo{IsTrans}(logp; kwargs...) = SimpleVarInfo{IsTrans}(kwargs, logp)`
- `SimpleVarInfo(values, logp) = SimpleVarInfo{false}(values, logp)`
- `SimpleVarInfo(logp; kwargs...) = SimpleVarInfo{false}(logp; kwargs...)`

And possible something with default values of `logp`. And/or `SimpleVarInfo{IsTrans,T}(values)`. But then again it becomes confusing :smile: 

So maybe for a start just define some of them, and then add more later if desired.

yebai replied:
I like the `SimpleVarInfo{IsTrans}(values, logp)` style better than alternatives.

yebai replied:
Probably too many constructors here - it is a bit confusing.

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -178,8 +199,10 @@ function acclogp!!(vi::SimpleVarInfo{<:Any,<:Ref}, logp)
+function Base.show(
+    io::IO, ::MIME"text/plain", svi::SimpleVarInfo{<:Any,<:Any,IsTrans}
+) where {IsTrans}
+    return print(io, "SimpleVarInfo{IsTrans=$(IsTrans)}(", svi.values, ", ", svi.logp, ")")
```
IMO this should print a proper constructor at least.

torfjelde replied:
Yeah I wasn't quite certain about this. Immediately I was thinking of using `SimpleVarInfo(svi.values, svi.logp, IsTrans)` but that seemed a bit misleading.

So I guess we have a couple of options:
1. `SimpleVarInfo(svi.values, svi.logp, IsTrans)`: simple and nice looking but is a bit misleading in how it represents the instance.
2. `SimpleVarInfo(svi.values, svi.logp)`: simplest but doesn't show `IsTrans` which is very important information.
3. `SimpleVarInfo{...}(svi.values, svi.logp)`: the default `show` which shows the entire type, etc.

I _think_ I'm in favour of  (1), despite it being somewhat misleading.

yebai replied:
Option 1 looks reasonalbe - can we print something along the lines `Unconstrained SimpleVarInfo(svi.values, svi.logp)` and `Constrained SimpleVarInfo(svi.values, svi.logp)` here? We don't need to stick to valid Julia syntax since it is just a  string for display.

torfjelde commented on src/context_implementations.jl:
```diff
@@ -462,19 +474,24 @@ function get_and_set_val!(
-            push!!(vi, vn, r[:, i], dist, spl)
-            settrans!(vi, false, vn)
+            if istrans(vi)
+                push!!(vi, vn, maybe_link(vi, vn, dist, r[:, i]), dist, spl)
```
Should I maybe make this
```suggestion
                push!!(vi, vn, link(vi, vn, dist, r[:, i]), dist, spl)
```
instead to make it clear that we're definitively linking rather than just "maybe"?

Also, I just realized that with the current implementation of `VarInfo`, this part of the `if` statement is _never_ hit because `istrans(vi)` is always `false`. This also means that it isn't tested (though it should work just fine), so maaaybe we should just remove this?

yebai replied:
Sounds good to remove the `maybe_` prefix - I think `link` is sufficiently clear since it is a common practice to overload method behaviour using multi-dispatching.

yebai commented:
> The main question is whether we want to just make getindex_raw be getindex, making it a breaking change, or if we want to keep getindex_raw for now, make a non-breaking change, and then we make a breaking change at a later stage.


Let's keep it `getindex_raw` for now - we would like a few more incremental PRs to unify the `VarInfo` and `SimpleVarInfo` terminology. It's better to keep things explicit to avoid confusion during the transition period.

yebai commented on src/simple_varinfo.jl:
```diff
@@ -126,14 +126,36 @@ ERROR: type NamedTuple has no field b
+function SimpleVarInfo(values, logp, istrans::Bool)
+    return SimpleVarInfo(values, logp, Val{istrans}())
+end
+function SimpleVarInfo(values, logp, ::Val{IsTrans}) where {IsTrans}
```
Maybe introduced a type to improve clarity, e.g.: 

```julia
abstract type Parameters end

struct Constrained <: Parameters end
struct Unconstrained <: Parameters end

function SimpleVarInfo(values, logp, T) where {T<:Parameters} end
```

yebai commented:
> feel about exporting evaluate!!? 

It sounds good to export `evaluate!!`. We also want to consider slowly unifying model operations using the same `evaluate!!` function. For example, `evaluate!!(m, vi, ctx_sampling)` or `evaluate!!(m, vi, ctx_logdensity)`.

yebai commented:
bors try

yebai commented on src/context_implementations.jl:
```diff
@@ -194,7 +194,9 @@ end
+    # x = vi[vn]
+    r_raw = getindex_raw(vi, vn)
+    r = maybe_invlink(vi, vn, dist, r_raw)
     return r, Bijectors.logpdf_with_trans(dist, r, istrans(vi, vn)), vi
```
it'll be nice if we can customise the transform used inside `logpdf_with_trans` in the future.

torfjelde replied:
We can actually replace the above with

``` julia
logp = zero(eltype(vi))
if istrans(vi, vn)
    r, logjac = forward(bijector(dist))
    logp -= logjac
else
    r = r_raw
end

return r, logpdf(dist, r)
```

right now. _But_ I think if we do this we should remove `logpdf_with_trans` everywhere, so I wanted to defer that to another PR.

yebai replied:
@phipsgabler this is what I mentioned today. We can now replace `logpdf_with_trans` with ChangeOfVariables/LogDensityInterface API. 

[#logdensity](https://github.com/TuringLang/DynamicPPL.jl/pull/342)

devmotion commented:
> How do you @yebai @devmotion feel about exporting `evaluate!!`? I'm referring to it in some docstrings, and given that it's a bit more pleasant to work with now than it was in the old days, maybe exposing this to the user (of DPPL granted, not of Turing) isn't the worst idea?

Yes, seems reasonable. I remember that it was a bit of a mess with the different function signatures but maybe this was only the problem with `(model::Model)(...)`? If this is (still) the case, it might be good to clean it a bit before exporting `evaluate!!` to reduce the number of breaking releases (as it would be part of the official API). That's my only suggestion though, and it doesn't matter much if we don't care about breaking releases.

yebai commented:
@torfjelde if the tests are passing locally, please feel free to trigger bors to merge this PR.

torfjelde commented:
bors r+

torfjelde commented:
bors r+

devmotion commented:
> > How do you @yebai @devmotion feel about exporting `evaluate!!`? I'm referring to it in some docstrings, and given that it's a bit more pleasant to work with now than it was in the old days, maybe exposing this to the user (of DPPL granted, not of Turing) isn't the worst idea?
> 
> Yes, seems reasonable. I remember that it was a bit of a mess with the different function signatures but maybe this was only the problem with `(model::Model)(...)`? If this is (still) the case, it might be good to clean it a bit before exporting `evaluate!!` to reduce the number of breaking releases (as it would be part of the official API). That's my only suggestion though, and it doesn't matter much if we don't care about breaking releases.

@torfjelde It seems the dispatches for `evaluate!!` are indeed a bit messy still: https://github.com/TuringLang/DynamicPPL.jl/issues/370 Maybe we should really clean the dispatch pipeline and possibly remove some function signatures before exporting it?

torfjelde commented:
Yeah, let's do that :+1:

torfjelde commented:
> Maybe the API should mainly be evaluate!!(model, varinfo, context)? The dispatch problems arise from us trying to support samplers, rngs, and different combinations with and without varinfo and context as well.

So I agree with this :+1: My point is that if we just straight up remove methods such as `(model::Model)(rng)` (this isn't directly implemented, but it's _effectively_ implemented since we have an `evaluate!!` for it), etc., we're going to break a lot of downstream code. So should we move these "bad" signatures to `(model::Model)(...)` + deprecate them, and then we only have `evaluate!!(model, varinfo, sampler)`?

torfjelde commented:
Sweet, then we're on the same page :+1: Have made those changes locally; running tests as we speak.

devmotion commented:
If it's too breaking for downstream code (even though it isn't technically breaking since `evaluate!!` is not exported) we could move the `evaluate!!`/`(::Model)` changes to a separate breaking PR. The only question then would be if we want to export `evaluate!!` in this PR already or if it should be done in the other PR.

torfjelde commented:
bors r+

yebai commented:
>  then we only have `evaluate!!(model, varinfo, sampler)`?

We should probably aim for a more flexible API which does not depend on samplers, i.e. `evaluate!!(model, varinfo context)`

torfjelde commented:
Aight, so I've hit a bit of a snag with this PR. There's a bug in how `SimpleVarInfo` handles broadcasting when LHS is a `Matrix`. It can be boiled down to the following:

``` julia
julia> @model function demo(N=3)
           x = Array{Real}(undef, N, N)
           x .~ MvNormal(zeros(N), I)
       end
demo (generic function with 3 methods)

julia> m = demo();

julia> vi = VarInfo(model);

julia> vns = [@varname(x[:,1]), @varname(x[:,2]), @varname(x[:,3])];

julia> vi[vns] # returns `Matrix`, which TBH is weird
3×3 Matrix{Float64}:
  0.667867  -0.967538   1.07314
 -1.47863    0.174372   0.970644
 -0.950455  -2.58733   -0.598436

julia> svi[vns] # returns `Vector{<:Vector}` as expected
3-element Vector{Vector{Float64}}:
 [-0.4991167883862232, 0.12240684039735461, -1.5039083397003628]
 [0.1350029260313995, -0.3719874023591045, 0.5128746327707455]
 [0.6553486495480079, 0.06209101530919303, 0.2660959301320175]
```

To address this I think we have a couple of alternatives.
1. Just ignore this issue for now, simply overloading this case for `SimpleVarInfo` to `error`.
2. Alter `Base.getindex(::VarInfo, vns)`.


For (2) I'm thinking something along the lines of

``` julia
# For `VarInfo` (we also have to do this for `vns::VarName`, etc. but you get the idea)
function getindex(vi::VarInfo, vns::Vector{<:VarName}, dist::Distribution=getdist(vi, first(vns)))
    return if istrans(vi, vns[1])
        Bijectors.invlink(dist, reconstruct(dist, getval(vi, vns), length(vns)))
    else
        reconstruct(dist, getval(vi, vns), length(vns))
    end
end

# ...

# For `SimpleVarInfo`:
function Base.getindex(vi::SimpleVarInfo, vns::Vector{<:VarName}, dist::Distribution)
    return reconstruct(dist, reduce(vcat, vi[vns]), length(vns))
end
```

Then we have:

``` julia
julia> vi[vns, MvNormal(zeros(3), I)]
3×3 Matrix{Float64}:
  0.667867  -0.967538   1.07314
 -1.47863    0.174372   0.970644
 -0.950455  -2.58733   -0.598436

julia> svi[vns, MvNormal(zeros(3), I)]
3×3 Matrix{Float64}:
 -0.499117   0.135003  0.655349
  0.122407  -0.371987  0.062091
 -1.50391    0.512875  0.266096
```

This has the benefit that it preserves the current (somewhat strange) behavior of `VarInfo`, _but_ the drawback is that the API for `SimpleVarInfo` is inconsistent with `VarInfo`. I would argue that the current API for `VarInfo` is "wrong" anyways, but it still feels a bit weird.

We also don't _have_ to overload `getindex` (not sure how I feel about it myself, though I think I'm tiping towards "no").

yebai commented:
Let's go for the second option here, which is altering `Base.getindex(::VarInfo, vns)`. The API for `VarInfo` is not very consistent in its current form, so it is fine to prioritise `SimpleVarInfo` in this case.

yebai commented:
In the longer future, I think we might want to consider storing the shape information of LHS (e.g. `type(x)` and `size(x)`) inside `VarName` #373.

torfjelde commented:
> In the longer future, I think we might want to consider storing the shape information of LHS (e.g. type(x) and size(x)) inside VarName https://github.com/TuringLang/DynamicPPL.jl/issues/373.

I'm not a particularly big fan of this tbh :confused:  It'll work nicely with Distributions, but what about when we start returning `NamedTuple` from RHS, e.g. using MeasureTheory.jl? IMO we want to store as little information as possible in the `AbstractVarInfo` and make as much use of the information available to us at the `~` statement, as in this case of passing the `dist`. The more information we add into the `AbstractVarInfo` itself, the more complicated it's implementation will be.

_But_ I guess this is a discussion for another time:)

yebai commented:
@torfjelde Just a gentle reminder.

torfjelde commented:
I think this is now ready for a look-over!

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -1,11 +1,19 @@
+abstract type AbstractConstraint end
+
+struct Constrained <: AbstractConstraint end
+struct Unconstrained <: AbstractConstraint end
```
Maybe it would also be better to talk about whether the samples are transformed than whether they are unconstrained? In particular if we think about different possible transformations at some point.

devmotion replied:
Sure, just go ahead if you think it's a good idea 🙂

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -126,14 +188,26 @@ ERROR: type NamedTuple has no field b
     values::NT
+    "holds the accumulated log-probability"
     logp::T
+    "represents whether it assumes variables to be constrained or unconstrained"
+    constraint::C
```
I think this design might not be sufficient in the long run but possibly it's sufficient as a start. I guess in many scenarios we would like some variables to be stored in constrained and some to be stored unconstrained (i.e., transformed).

I'm not sure if this should be part of `SimpleVarInfo` at all. Maybe a context would be sufficient?

torfjelde replied:
We could always add this information to the type down the road though, no?

And regarding whether it should be attached to `SimpleVarInfo` or we should use a context instead, I think we'll have to defer for now as this will require fairly significant changes.

devmotion replied:
Sure, it would require more changes, which is why I think the approach in the PR is a good start maybe nevertheless. It just feels quite limiting that there is no more fine-grained control over which variables are transformed (and how) but everything has to be transformed or untransformed.

devmotion reviewed (commented):
I think it's a big step forward, in particular for `SimpleVarInfo`. I added some comments. More generally, I am not sure though if we should add something like `AbstractConstraint`. Maybe it would be better to make things more orthogonal and separate the transformation logic more clearly from `AbstractVarInfo`. Maybe eg we could add a context that allows to specify which variables are (supposed to be) transformed and which not? Possibly even specifying the type of transformation (some `DefaultTransformation`, a `NoTransformation`, and maybe even custom transformations) for individual and subsets of variables?

torfjelde commented on src/context_implementations.jl:
```diff
@@ -305,19 +310,22 @@ function dot_tilde_assume(
+
 function dot_tilde_assume(context::LikelihoodContext, right, left, vn, vi)
-    return dot_assume(NoDist.(right), left, vn, vi)
+    nodist = right isa Distribution ? NoDist(right) : NoDist.(right)
```
Yeah, I think it failed for the scenario where `NoDist` was wrapping something like a `MultivariateDistribution`, since in that case you're no longer hitting: https://github.com/JuliaStats/Distributions.jl/blob/7831193ecbb70967b9f1af71e1c8f0850b89f3a5/src/common.jl#L151

devmotion replied:
I see. Slightly annoying to have to write these branches in multiple places. Maybe we could define
```julia
nodist(d::Distribution) = NoDist(d)
```
and broadcast that instead?

torfjelde replied:
I like that!

torfjelde commented on src/context_implementations.jl:
```diff
@@ -412,19 +420,22 @@ function dot_assume(
-    dists::Union{Distribution,AbstractArray{<:Distribution}},
+    dist::Distribution, var::AbstractArray, vns::AbstractArray{<:VarName}, vi
+)
+    r = map(vn -> vi[vn, dist], vns)
+    lp = sum(Bijectors.logpdf_with_trans.(dist, r, map(Base.Fix1(istrans, vi), vns)))
```
The reason why I did it this way is because `Ref` destroys type-inference when used with Zygote (any usage of `Ref` will be treated as boxing `Any` IIRC), but maybe we shouldn't be tailoring our implementation to Zygote's  limitations?

But I guess as things are now, we generally recommend against using any form of `.~` if someone actually cares about performance.

devmotion replied:
> The reason why I did it this way is because `Ref` destroys type-inference when used with Zygote (any usage of `Ref` will be treated as boxing `Any` IIRC), but maybe we shouldn't be tailoring our implementation to Zygote's limitations?

Does that happen with `(vi,)` as well? That's an alternative to `Ref` that some people (and maybe Zygote?) prefer.

devmotion replied:
```suggestion
    r = map(vn -> vi[vn, dist], vns)
    lp = sum(Bijectors.logpdf_with_trans.(dist, r, istrans.((vi,), vns)))
```

torfjelde replied:
Ah, that's a good idea! We can use that:)

devmotion commented on src/context_implementations.jl:
```diff
@@ -438,7 +449,7 @@ function dot_assume(
     r = get_and_set_val!(rng, vi, vns, dists, spl)
     # Make sure `r` is not a matrix for multivariate distributions
-    lp = sum(Bijectors.logpdf_with_trans.(dists, r, istrans(vi, vns[1])))
+    lp = sum(Bijectors.logpdf_with_trans.(dists, r, map(Base.Fix1(istrans, vi), vns)))
```
Or
```suggestion
    lp = sum(Bijectors.logpdf_with_trans.(dist, r, istrans.((vi,), vns)))
```

torfjelde commented on src/context_implementations.jl:
```diff
@@ -412,19 +420,22 @@ function dot_assume(
-    # in which case `var` will have `undef` elements, even if `m` is present in `vi`.
-    r = reshape(vi[vec(vns)], size(vns))
-    lp = sum(Bijectors.logpdf_with_trans.(dists, r, istrans(vi, vns[1])))
+    @assert length(vns) == length(dists) == length(var)
```
Do we? Wouldn't something like `size(var) == (1, 2)` be valid even if `size(dists) == (2, )`?

torfjelde replied:
Maybe even the current `@assert` is too strong since it disallows something like `x .~ [Normal()]` unless `length(x) == 1`, which is inconsistent with what you'd expect from broadcast.

devmotion replied:
Well, that's what I was wondering :smile: So `length` is not sufficient since I guess we don't want e.g. `size(var) == (2, 3)` and `size(dists) == (3, 2)`. But that would be caught by the broadcasting machinery anyway, wouldn't it? So should we just remove the assertions? Or check that the shapes can be broadcasted (IIRC there's an unexported function in base for it)?

torfjelde replied:
I think let's just remove the `@assert` and let it error on subsequent broadcast if it's incorrect :shrug:

torfjelde replied:
I also replaced some of the `map` usages with the broadcasting using `(vi,)` because I now recall that I introduced the checks because otherwise stuff like `map(..., [@varname(x)], [Normal(), Normal()])` would just silently drop one of the `Normal()`.

torfjelde commented:
Want to give it another look now @devmotion ?:) Think I've addressed all your comments.

torfjelde commented:
bors won't be happy with this PR before https://github.com/TuringLang/Turing.jl/pull/1850 has been merged, so I'm going to point the #415 to this PR so we can just merge both in the same PR + make sure the tests are all passing

devmotion commented on src/distribution_wrappers.jl:
```diff
@@ -24,12 +27,20 @@ function Distributions.loglikelihood(dist::NamedDist, x::AbstractArray{<:Real})
 end
 NoDist(dist::NamedDist) = NamedDist(NoDist(dist.dist), dist.name)
 
+nodist(dist::Distribution) = NoDist(dist)
+nodist(dists::AbstractArray) = nodist.(dists)
```
However, maybe then it would be simpler to use `map` here:
```suggestion
nodist(dist::Distribution) = NoDist(dist)
nodist(dists::AbstractArray{<:Distribution}) = map(nodist, dists)
```

torfjelde replied:
> Usually not a fan of implicit broadcasting (since e.g. it breaks broadcast fusions) but I guess it's fine here (and we don't actually want broadcast fusion anywhere).

But isn't this why we introduced `nodist`? To avoid the `if` statement which is required because `nodist.(dists)` will break if `dists isa MultivariateDistribution`?

torfjelde replied:
Though I see what you're saying, is that really simpler than the original approach? We're not really going to be broadcast over `NoDist` in many scenarios, hence I don't think allowing fusion will make any difference whatsoever in practice.

devmotion replied:
Yeah, not sure how relevant it is in practice but it will surely reduce allocations whenever broadcast fusion is possible. I'd say, since we know about this possible optimization, it's not something that you're likely to discover otherwise, and the fix is simple, we should apply the optimization right away and ensure that we get rid of these allocations.

torfjelde replied:
Aighty then! Adding a `_protect_dists` :+1:

torfjelde replied:
`_protect_dists` doesn't work of course because it results in `(NoDist(dist), )` when `dist isa Distribution` when we actually want `NoDist(dist)`... 

So I've reverted to the `nodist` without fusion.

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -158,15 +232,16 @@ function BangBang.empty!!(vi::SimpleVarInfo)
 getlogp(vi::SimpleVarInfo) = vi.logp
-setlogp!!(vi::SimpleVarInfo, logp) = SimpleVarInfo(vi.values, logp)
-acclogp!!(vi::SimpleVarInfo, logp) = SimpleVarInfo(vi.values, getlogp(vi) + logp)
+setlogp!!(vi::SimpleVarInfo, logp) = Setfield.@set vi.logp = logp
+acclogp!!(vi::SimpleVarInfo, logp) = Setfield.@set vi.logp = getlogp(vi) + logp
```
Do we gain something from not just defining `... = SimpleVarInfo(vi.values, logp, vi.transformation)`? Seems slightly overkill here :shrug:

torfjelde replied:
So I just went with `@set` because at some point I forgot to update the method to also carry the `vi.transformation`, which lead to silent conversion to "unconstrained" `SimpleVarInfo`. So I figured I might as well just have `Setfield` handle this for me going forward to avoid such silly mistakes :shrug: It should be zero overhead anyways, right? Or are you worried about compile-times?

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -178,11 +253,29 @@ function acclogp!!(vi::SimpleVarInfo{<:Any,<:Ref}, logp)
+    return maybe_invlink(vi, vn, dist, Base.getindex(vi, vn))
+end
+function Base.getindex(vi::SimpleVarInfo, vns::Vector{<:VarName}, dist::Distribution)
+    vals_linked = map(vns) do vn
+        maybe_invlink(vi, vn, dist, Base.getindex(vi, vn))
+    end
+    return reconstruct(dist, reduce(vcat, vals_linked), length(vns))
```
Maybe use `mapreduce` and a recursive definition?
```suggestion
    vals_linked = mapreduce(vcat, vns) do vn
        getindex(vi, vn, dist)
    end
    return reconstruct(dist, vals_linked, length(vns))
```

torfjelde replied:
Ah yes, this I think is leftover from before I added the optional `dist` to `getindex`; thanks!

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -221,9 +314,23 @@ Base.getindex(vi::SimpleVarInfo, vns::Vector{<:VarName}) = map(Base.Fix1(getinde
+end
+getindex_raw(vi::SimpleVarInfo, vns::Vector{<:VarName}) = vi[vns]
+function getindex_raw(vi::SimpleVarInfo, vns::Vector{<:VarName}, dist::Distribution)
+    vals = getindex_raw(vi, vns)
+    # `reconstruct` expects a flattened `Vector` regardless of the type of `dist`, so we `vcat` everything.
+    return reconstruct(dist, reduce(vcat, vals), length(vns))
```
Yeah, a bit annoying that we have to `vcat`. Since `getindex` calls `map` anyway (I think?), maybe it would be more efficient again to use `mapreduce`:
```suggestion
    # `reconstruct` expects a flattened `Vector` regardless of the type of `dist`, so we `vcat` everything.
    vals = mapreduce(Base.Fix1(getindex_raw, vi), vcat, vns)
    return reconstruct(dist, vals, length(vns))
```

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -337,79 +444,90 @@ function Base.eltype(
+    vi = BangBang.setindex!!(vi, value_raw, vns)
+
+    # Compute logp.
+    lp = sum(Bijectors.logpdf_with_trans.(dists, value, map(Base.Fix1(istrans, vi), vns)))
```
```suggestion
    lp = sum(Bijectors.logpdf_with_trans.(dists, value, istrans.((vi,), vns)))
```

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -337,79 +444,90 @@ function Base.eltype(
-    value = f.(vns, dists)
-    vi = BangBang.setindex!!(vi, value, vns)
-    lp = sum(Distributions.logpdf.(dists, value))
+    @assert length(dist) == size(var, 1)
```
Maybe move this down to just before `zip` (and check `length(vns) == size(value, 2)`) so it's clearer why there is an assertion. Or maybe even make it a proper error.

torfjelde replied:
I've added error message:)

torfjelde commented on src/context_implementations.jl:
```diff
@@ -496,12 +508,13 @@ function get_and_set_val!(
-            r = reshape(vi[vec(vns)], size(vns))
+            # r = reshape(vi[vec(vns)], size(vns))
+            r_raw = getindex_raw(vi, vec(vns))
+            r = maybe_invlink.(Ref(vi), vns, dists, reshape(r_raw, size(vns)))
```
Yes it would, but it seems like this will require some work (atm it's not 100% clear to me how to handle this in full generality) and so I didn't want to start messing around with it.

torfjelde commented on src/simple_varinfo.jl:
```diff
@@ -178,11 +253,29 @@ function acclogp!!(vi::SimpleVarInfo{<:Any,<:Ref}, logp)
 end
 
-function Base.show(io::IO, ::MIME"text/plain", svi::SimpleVarInfo)
+function Base.show(
+    io::IO, ::MIME"text/plain", svi::SimpleVarInfo{<:Any,<:Any,<:NoTransformation}
+)
     return print(io, "SimpleVarInfo(", svi.values, ", ", svi.logp, ")")
 end
 
+function Base.show(
+    io::IO, ::MIME"text/plain", svi::SimpleVarInfo{<:Any,<:Any,<:DefaultTransformation}
+)
+    return print(io, "Transformed SimpleVarInfo(", svi.values, ", ", svi.logp, ")")
+end
+
```
Sure! Is there a particular reason why we don't do `!(vi.transformation isa NoTransformation)`?

devmotion replied:
No, that shouldn't matter.

torfjelde commented:
bump @devmotion  :eyes:

devmotion commented on src/simple_varinfo.jl:
```diff
@@ -337,79 +440,90 @@ function Base.eltype(
+    value = f.(vns, dists)
+
+    # Transform if we're working in transformed space.
+    value_raw = if dists isa Distribution
+        maybe_link.((vi,), vns, (dists,), value)
+    else
+        maybe_link.((vi,), vns, dists, value)
+    end
```
This is where I assume something like
```julia
_protect_dist(d::Distribution) = (d,)
_protect_dist(x) = x
```
could be useful:
```suggestion
    value_raw = maybe_link.((vi,), vns, _protect_dist(dists), value)
```

But if this is the only place where it would be used, a separate function seems unnecessary.

devmotion reviewed (approved):
Sorry for the slow response, I'm on vacation until beginning of August.

The PR looks good to me :+1:

torfjelde commented:
> Sorry for the slow response, I'm on vacation until beginning of August.

Ah sorry, wasn't aware of that! Enjoy it man:)

torfjelde commented:
bors r+

torfjelde commented:
bors r+

yebai commented:
bors r+

torfjelde commented:
bors r+

torfjelde commented:
Unfortunately integration tests will fail because the tests for `ESS` are making assumptions about the `DEMO_MODELS` that are no longer true. As those are indeed the only tests that are failing, hence I suggest we just merge this and then update the tests in Turing accordingly (I have a working update locally).

In addition, I've bumped the minimum supported Julia version to 1.6 for DPPL; 1.6 is now LTS and we've already done this for Turing.jl, so continue to test against 1.3 but _not_ 1.6 seems strange.

torfjelde commented:
bors r+

torfjelde commented:
@yebai Can you just merge this once Bors comes back with failure, assuming the ESS tests are the only ones failing?

torfjelde commented:
Actually, given that we bump the lower bound for the Julia version, we should bump minor rather than patch version. This will also make bors happy :+1:

torfjelde commented:
bors r+
